portability-worker/src/main/java/org/dataportabilityproject/worker/WorkerMain.java

```diff
@@ -18,17 +18,66 @@
 import com.google.common.util.concurrent.UncaughtExceptionHandlers;
 import com.google.inject.Guice;
 import com.google.inject.Injector;
+import org.dataportabilityproject.api.launcher.ExtensionContext;
+import org.dataportabilityproject.api.launcher.Logger;
+import org.dataportabilityproject.spi.transfer.extension.TransferExtension;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
 
 /**
  * Main class to bootstrap a portabilty worker that will operate on a single job whose state
  * is held in WorkerJobMetadata.
  */
 public class WorkerMain {
-  public static void main(String[] args) throws Exception {
-    Thread.setDefaultUncaughtExceptionHandler(UncaughtExceptionHandlers.systemExit());
-    Injector injector = Guice.createInjector(new WorkerModule());
-    WorkerImpl worker = injector.getInstance(WorkerImpl.class);
-    worker.processJob();
-    System.exit(0);
-  }
+    public static void main(String[] args) throws Exception {
+        Thread.setDefaultUncaughtExceptionHandler(UncaughtExceptionHandlers.systemExit());
+
+        ExtensionContext extensionContext = getExtensionContext();
+
+        List<TransferExtension> extensions = new ArrayList<>();
+        ServiceLoader.load(TransferExtension.class).iterator().forEachRemaining(extensions::add);
+
+        for (TransferExtension extension : extensions) {
+            extension.initialize(extensionContext);
+        }
+        for (TransferExtension extension : extensions) {
+            extension.getExporters();
+            extension.getImporters();
+        }
+        for (TransferExtension extension : extensions) {
+            extension.start();
+        }
+
+        Injector injector = Guice.createInjector(new WorkerModule());
+        WorkerImpl worker = injector.getInstance(WorkerImpl.class);
+        worker.processJob();
+
+        for (TransferExtension extension : extensions) {
+            extension.shutdown();
+        }
+
+        System.exit(0);
+    }
+
+    private static ExtensionContext getExtensionContext() {
+        return new ExtensionContext() {
+                @Override
+                public Logger getLogger() {
+                    return new Logger() {{
```

Comment:
Does this mean there is a logger per extension?

Reply:
Ah that's a good point. Maybe we want to change that to getLogger(Object key)?

Then, the launcher implementation is free to decide what to return.

Reply:
Yes, that sounds good.